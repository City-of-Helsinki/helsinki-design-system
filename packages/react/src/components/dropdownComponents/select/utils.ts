import React from 'react';

import { SelectData, SelectMetaData, SelectDataHandlers, FilterFunction } from './types';
import {
  Group,
  Option,
  ModularOptionListData,
  ModularOptionListProps,
  ScreenReaderNotification,
} from '../modularOptionList/types';
import { ChangeEvent } from '../../dataProvider/DataContext';
import { eventTypes } from './events';
import {
  iterateAndCopyGroup,
  getGroupLabelOption,
  getAllOptions,
  propsToGroups,
  getSelectedOptions,
} from '../modularOptionList/utils';

// Re-export commonly used utility functions
export {
  getAllOptions,
  getSelectedOptions,
  propsToGroups,
  updateGroupLabelAndOptions,
  updateOptionInGroup,
  defaultFilter,
} from '../modularOptionList/utils';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

export function filterOptions(groups: ModularOptionListData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function filterSelectableOptions(
  groups: ModularOptionListData['groups'],
  filterStr: string,
  filterFunc: FilterFunction,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).filter((option) => filterFunc(option, filterStr));
}

export function mergeSearchResultsToCurrent(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
  currentGroups: ModularOptionListData['groups'],
): ModularOptionListData['groups'] {
  const newData = propsToGroups(props) || [];
  const newOptions = getAllOptions(newData);
  const currentOptionsWithoutMatches = getSelectedOptions(currentGroups).filter((option) => {
    const sameInNewOptionsIndex = newOptions.findIndex((newOption) => {
      return newOption.value === option.value;
    });
    if (sameInNewOptionsIndex > -1) {
      newOptions[sameInNewOptionsIndex].selected = true;
      return false;
    }
    return true;
  });

  const currentHiddenOptionsInAGroup = currentOptionsWithoutMatches.length
    ? [{ options: currentOptionsWithoutMatches.map((opt) => ({ ...opt, visible: false })) } as Group]
    : [];

  return [...currentHiddenOptionsInAGroup, ...newData];
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}
