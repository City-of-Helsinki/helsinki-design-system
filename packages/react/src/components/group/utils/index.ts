import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
