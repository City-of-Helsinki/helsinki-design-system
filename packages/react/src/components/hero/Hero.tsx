/* eslint-disable prettier/prettier */
import React from 'react';

import '../../styles/base.module.css';
import styles from './Hero.module.scss';
import classNames from '../../utils/classNames';
import { useTheme } from '../../hooks/useTheme';
import { getShapeHeight, Koros, KorosProps } from '../koros';
import { Text } from './Text';
import { Title } from './Title';
import { IconArrowDown } from '../../icons';

type HTMLElementAttributes = React.HtmlHTMLAttributes<HTMLDivElement>;
type ImgElementAttributes = React.ImgHTMLAttributes<HTMLImageElement>;
export type HeroProps = React.PropsWithChildren<
  HTMLElementAttributes & {
    /**
     * Show arrow icon
     */
    showArrowIcon?: boolean;
    /**
     * Should texts be centered. Use when there is no image.
     */
    centeredContent?: boolean;
    /**
     * Image source
     */
    imageSrc?: string;
    /**
     * Additional information seen below the hero. Usually photographer information. The element has aria-hidden like the img element.
     */
    information?: string;
    /**
     * Koros properties. Accepts also boolean "hide", which hides the koros.
     * @see KorosProps
     */
    koros?: Omit<KorosProps, 'rotate'> & {
      hide?: boolean;
    };
    /*
     * Css variables for theming the hero
     */
    theme?: HeroCustomTheme;
    /**
     * Defines the hero variant
     */
    variant?: 'imageLeft' | 'imageRight' | 'backgroundImage' | 'imageBottom' | 'diagonalKoros' | 'noImage';
  }
>;

export interface HeroCustomTheme {
  /**
   * Arrow icon color
   * Default --color
   */
  '--arrow-icon-color'?: string;
  /**
   * Sets the background color. Is also koros color, if --koros-color is not set.
   * Default --color-white
   */
  '--background-color'?: string;
  /**
   * Background color of the bottom area with arrow and information
   * Default transparent
   */
  '--bottom-background-color'?: string;
  /**
   * Text color
   * Default --color-black-90
   */
  '--color'?: string;
  /**
   * How background image is positioned if it overflows the container. Is set to css property "object-position"
   * Default bottom right
   */
  '--image-position'?: string;
  /**
   * Color of the koros
   * Default --background-color
   */
  '--koros-color'?: string;
  /**
   * Height of the koros
   * Default 85px
   */
  '--koros-height'?: string;
  /**
   * The "diagonalKoros" variant might need koros position adjustment, if texts overflow.
   * This set the koros element's css rule "right"
   * Default "45%"
   */
  '--diagonal-koros-position'?: string;
  /**
   * Custom horizontal padding for small screens
   * Default --spacing-layout-2-xs
   */
  '--horizontal-padding-small'?: string;
  /**
   * Custom horizontal padding for medium screens
   * Default --spacing-layout-xs
   */
  '--horizontal-padding-medium'?: string;
  /**
   * Custom horizontal padding for large screens
   * Default --spacing-layout-xs
   */
  '--horizontal-padding-large'?: string;
  /**
   * Custom horizontal padding for x-large screens
   * Default --spacing-layout-xs
   */
  '--horizontal-padding-x-large'?: string;
  /**
   * Text color of the information text.
   * Default inherit
   */
  '--information-color'?: string;
  /**
   * Extra top padding for the information element. Has no effect when Hero variant is 'imageLeft' or 'imageRight'.
   * Default 0
   */
  '--information-padding-top'?: string;
}

export const Hero = ({
  children,
  variant,
  imageSrc,
  showArrowIcon,
  centeredContent,
  information,
  theme,
  koros,
  ...elementAttributes
}: HeroProps) => {
  const editableTheme = { ...theme };
  if (!editableTheme['--koros-color']) {
    editableTheme['--koros-color'] = 'var(--background-color)';
  }
  if (variant === 'backgroundImage') {
    editableTheme['--top-koros-color'] = 'var(--background-color)';
    editableTheme['--bottom-koros-color'] = 'var(--koros-color)';
  }
  if (!editableTheme['--koros-height']) {
    const korosHeight = getShapeHeight(koros || {});
    if (korosHeight) {
      editableTheme['--koros-height'] = `${korosHeight}px`;
    }
  }

  const currentVariant: HeroProps['variant'] = variant || (imageSrc ? 'imageLeft' : 'noImage');
  const customThemeClass = useTheme<HeroCustomTheme>(styles.hero, editableTheme);
  const korosStyle = { ...koros, style: { fill: 'var(--koros-color)' } };

  const heroElementAttributes: HTMLElementAttributes = {
    ...elementAttributes,
    className: classNames(
      styles.hero,
      customThemeClass,
      styles[currentVariant],
      (elementAttributes as HTMLElementAttributes).className,
    ),
  };

  const ContentWithArrowIconAndInfoWrapper = ({ children: innerChildren }: { children: React.ReactNode }) => {
    return (
      <>
        {innerChildren}
        <div className={styles.bottomContainer} aria-hidden="true">
          <div className={styles.bottomContentAligner}>
            {showArrowIcon && (
              <div className={styles.arrowIconContainer}>
                <IconArrowDown className={styles.arrowIcon} />
              </div>
            )}
            {information && (
              <div className={classNames(styles.informationContainer)}>
                <span>{information}</span>
              </div>
            )}
          </div>
        </div>
      </>
    );
  };

  const Image = () => {
    return (
      /* eslint-disable-next-line jsx-a11y/alt-text */
      <img className={styles.image} src={imageSrc} aria-hidden alt="" />
    );
  };

  const ImageAsBackground = (props?: Pick<ImgElementAttributes, 'className'>) => {
    return (
      <div className={classNames(styles.withBackgroundBackground, props && props.className)}>
        <Image />
      </div>
    );
  };

  const TwoColumsImage = () => {
    return (
      <div key={`imageContainer-${currentVariant}`} className={styles.twoColumnsImageContainer}>
        <Image />
      </div>
    );
  };

  const Content = () => {
    const classNameList = centeredContent ? classNames(styles.card, styles.centeredContent) : classNames(styles.card);
    return <div className={classNameList}>{children}</div>;
  };

  const KorosInContainer = (
    props: KorosProps & {
      inward?: boolean;
      containerClassName?: string;
    },
  ) => {
    const { inward, containerClassName, style, ...korosProps } = props;
    const className =
      containerClassName || (inward !== true ? styles.korosContainer : styles.korosContainerInwardKoros);
    return (
      <div className={className}>
        <Koros {...{ ...korosProps, style, shift: false, compact: false }} />
      </div>
    );
  };

  if (currentVariant === 'backgroundImage') {
    return (
      <div {...heroElementAttributes}>
        <ContentWithArrowIconAndInfoWrapper>
          <div className={styles.withBackgroundContainer}>
            <div className={classNames(styles.withBackgroundBackground)}>
              <Image />
              <KorosInContainer {...korosStyle} containerClassName={styles.backgroundImageKoros} />
            </div>
            <div className={classNames(styles.content)}>
              <Content />
              <div className={styles.emptyColumn} />
            </div>
          </div>
        </ContentWithArrowIconAndInfoWrapper>
      </div>
    );
  }

  if (currentVariant === 'diagonalKoros') {
    return (
      <div {...heroElementAttributes}>
        <ContentWithArrowIconAndInfoWrapper>
          <div className={styles.diagonalKorosWithBackgroundContainer}>
            <div className={styles.content}>
              <div className={styles.contentColums}>
                <Content />
                <div className={styles.emptyColumn} />
              </div>
            </div>
            <div className={styles.korosAligner}>
              <div className={styles.diagonalKorosAndBackground}>
                <KorosInContainer {...korosStyle} />
              </div>
            </div>
            <ImageAsBackground className={styles.diagonalKorosBackgroundContainer} />
          </div>
        </ContentWithArrowIconAndInfoWrapper>
      </div>
    );
  }

  const hideKoros = !!koros?.hide;
  const flipVertical = koros?.flipVertical;
  const hasImage = !!imageSrc && currentVariant !== 'noImage';
  // const hasBottomArea = !!showArrowIcon || !!information;
  const columnStyle = hasImage && currentVariant !== 'imageBottom' ? styles.twoColumns : styles.singleColumn;
  return (
    <div {...heroElementAttributes}>
      <ContentWithArrowIconAndInfoWrapper>
        <div className={styles.container}>
          <div key="content" className={classNames(styles.content, columnStyle)}>
            {hasImage && currentVariant === 'imageLeft' && <TwoColumsImage />}
            <Content />
            {hasImage && currentVariant === 'imageRight' && <TwoColumsImage />}
          </div>
        </div>
        {!hasImage && !hideKoros && (
          <div key="korosAndImageContainer" className={classNames(styles.korosAndImageContainer)}>
            <KorosInContainer
              {...korosStyle}
              inward={!flipVertical}
              flipVertical={flipVertical}
              containerClassName={flipVertical ? styles.korosContainerOverflowBottom : undefined}
            />
          </div>
        )}
        {/*! hasImage && !hideKoros && hasBottomArea && (
          <div key="korosAndImageContainer" className={classNames(styles.korosAndImageContainer)}>
            <KorosInContainer {...korosStyle} flipVertical={flipVertical} inward={!flipVertical} />
          </div>
        ) */}
        {hasImage && (
          <div key="korosAndImageContainer" className={classNames(styles.korosAndImageContainer)}>
            <KorosInContainer {...korosStyle} flipVertical={flipVertical !== false} />
            <div key="imageContainer" className={classNames(styles.imageBelowKoros)}>
              <Image />
            </div>
          </div>
        )}
      </ContentWithArrowIconAndInfoWrapper>
    </div>
  );
};

Hero.Title = Title;
Hero.Text = Text;
