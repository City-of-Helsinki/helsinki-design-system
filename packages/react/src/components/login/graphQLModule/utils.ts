import { QueryOptions } from '@apollo/client';
import { merge } from 'lodash';

import { GraphQLModuleModuleProps } from '.';
import { ApiTokenClient } from '../apiTokensClient';

type PartialContext = Partial<QueryOptions['context']>;
type QueryHeaders = Record<string, unknown>;

function mergeQueryOptionsContexts(target: PartialContext, overrides: PartialContext): PartialContext {
  return merge(target, overrides);
}

export function mergeQueryOptionContexts(options: QueryOptions, overrideContext: PartialContext): QueryOptions {
  const context: PartialContext = options.context || {};
  // eslint-disable-next-line no-param-reassign
  options.context = mergeQueryOptionsContexts(context, overrideContext);
  return options;
}

export function mergeHeadersToQueryOptions(options: QueryOptions, headers: QueryHeaders): QueryOptions {
  const context: PartialContext = {
    headers,
  };
  return mergeQueryOptionContexts(options, context);
}

export function mergeAuthorizationHeaderToQueryOptions(options: QueryOptions, authHeader: string): QueryOptions {
  const headers: QueryHeaders = {
    authorization: authHeader,
  };
  return mergeHeadersToQueryOptions(options, headers);
}

export function mergeQueryOptions(
  target: Partial<GraphQLModuleModuleProps['queryOptions']>,
  overrides: Partial<GraphQLModuleModuleProps['queryOptions']>,
) {
  return merge(target, overrides);
}

/* No need to export to hds-react package */
export function mergeQueryOptionsToModuleProps<T, Q>(
  target: Partial<GraphQLModuleModuleProps<T, Q>>,
  queryOptions: Partial<QueryOptions<Q>>,
) {
  const { query, ...rest } = queryOptions;
  // eslint-disable-next-line no-param-reassign
  target.queryOptions = mergeQueryOptions(target.queryOptions || {}, rest);
  if (query) {
    // eslint-disable-next-line no-param-reassign
    target.query = query;
  }
  return target;
}

export function setBearerToQueryOptions(queryOptions: QueryOptions, token: string) {
  if (!token) {
    return queryOptions;
  }
  return mergeAuthorizationHeaderToQueryOptions(queryOptions, `Bearer ${token}`);
}

export function appendFetchOptions(queryOptions: QueryOptions, newOptions: Partial<Parameters<typeof fetch>[1]>) {
  if (!queryOptions.context) {
    // eslint-disable-next-line no-param-reassign
    queryOptions.context = {};
  }

  // eslint-disable-next-line no-param-reassign
  queryOptions.context.fetchOptions = {
    ...queryOptions.context.fetchOptions,
    ...newOptions,
  };
  return queryOptions;
}

/* No need to export to hds-react package */
export function mergeQueryOptionModifiers({
  options,
  queryHelper,
}: Pick<GraphQLModuleModuleProps, 'queryHelper' | 'options'>): Required<GraphQLModuleModuleProps>['queryHelper'] {
  const { apiTokenKey } = options || {};
  if (!apiTokenKey && !queryHelper) {
    return (opt: QueryOptions) => opt;
  }

  if (!apiTokenKey && queryHelper) {
    return queryHelper;
  }

  const tokenPicker = (apiTokenClient: ApiTokenClient, key: string) => {
    if (!apiTokenClient) {
      return undefined;
    }
    const tokens = apiTokenClient.getTokens();
    return tokens ? tokens[key] : undefined;
  };

  const apiTokenSetter: GraphQLModuleModuleProps['queryHelper'] = (opt, apiTokenClient) => {
    if (!apiTokenKey || !apiTokenClient) {
      return opt;
    }
    const token = tokenPicker(apiTokenClient, apiTokenKey);
    if (token) {
      return setBearerToQueryOptions(opt, token);
    }
    return opt;
  };

  if (!queryHelper) {
    return apiTokenSetter;
  }

  return (opt, apiTokenClient, beacon) => {
    return queryHelper(apiTokenSetter(opt, apiTokenClient, beacon), apiTokenClient, beacon);
  };
}
