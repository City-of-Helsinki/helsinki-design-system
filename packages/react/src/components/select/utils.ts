import React, { ReactElement, ReactNode } from 'react';

import {
  SelectData,
  Group,
  SelectProps,
  Option,
  OptionInProps,
  SelectMetaData,
  GroupInProps,
  FilterFunction,
} from './types';
import { getChildrenAsArray } from '../../utils/getChildren';
import { ChangeEvent } from '../dataProvider/DataContext';
import { eventTypes } from './events';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getOptionGroupIndex(groups: SelectData['groups'], option: OptionInProps): number {
  if (groups.length === 0) {
    return -1;
  }
  if (groups.length === 1) {
    return 0;
  }
  return groups.findIndex(({ options }) => {
    return (
      options.findIndex(({ value, isGroupLabel }) => value === option.value && isGroupLabel === option.isGroupLabel) >
      -1
    );
  });
}

export function iterateAndCopyGroup(groups: Group[], iterator: OptionIterator): Group[] {
  return groups.map((group, groupIndex) => {
    return {
      options: group.options.map((opt, optionIndex) => {
        return iterator(opt, group, optionIndex, groupIndex) || { ...opt };
      }),
    };
  });
}

export function getSelectedOptionsPerc(group: Group, pendingSelectionCount = 0): number {
  const optionCountWithoutGroupLabel = group.options.length - 1;
  if (!optionCountWithoutGroupLabel) {
    return 0;
  }
  return (
    (group.options.filter((option) => !option.isGroupLabel && option.selected).length + pendingSelectionCount) /
    optionCountWithoutGroupLabel
  );
}

export function updateOptionInGroup(
  groups: SelectData['groups'],
  updatedOption: Option,
  isMultiSelect: boolean,
): SelectData['groups'] {
  if (updatedOption.isGroupLabel) {
    throw new Error('Use updateGroupLabelAndOptions to update groupLabel and its related options');
  }
  const groupIndex = getOptionGroupIndex(groups, updatedOption);
  return groups.map((group, index) => {
    const selectedOptionPerc =
      index === groupIndex && isMultiSelect ? getSelectedOptionsPerc(group, updatedOption.selected ? 1 : -1) : 0;
    return {
      options: group.options.map((option) => {
        if (option.isGroupLabel) {
          if (index === groupIndex && isMultiSelect) {
            return {
              ...option,
              selected: selectedOptionPerc === 1,
            };
          }
          return { ...option };
        }
        if (index === groupIndex && option.value === updatedOption.value) {
          return {
            ...updatedOption,
            selected: !!updatedOption.selected,
          };
        }
        return {
          ...option,
          selected: !isMultiSelect ? false : option.selected,
        };
      }),
    };
  });
}

export function updateGroupLabelAndOptions(groups: SelectData['groups'], updatedOption: Option): SelectData['groups'] {
  if (!updatedOption.isGroupLabel) {
    throw new Error('Use updateOptionInGroup to update non-groupLabel options.');
  }
  if (!updatedOption.visible) {
    throw new Error('Cannot click an group label that is not visible (without a label)');
  }
  const targetGroupIndex = getOptionGroupIndex(groups, updatedOption);
  if (targetGroupIndex < 0) {
    return groups;
  }

  return iterateAndCopyGroup(groups, (option, group, optionIndex, groupIndex) => {
    if (groupIndex !== targetGroupIndex) {
      return { ...option };
    }
    return option.visible && !option.disabled
      ? {
          ...option,
          selected: updatedOption.selected,
        }
      : { ...option };
  });
}

export function clearAllSelectedOptions(groups: SelectData['groups']): SelectData['groups'] {
  return iterateAndCopyGroup(groups, (option) => {
    return {
      ...option,
      selected: false,
    };
  });
}

export function getAllOptions(groups: SelectData['groups'], filterOutGroupLabels = true): Option[] {
  const options: Option[] = [];
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (filterOutGroupLabels && option.isGroupLabel) {
        return;
      }
      options.push(option);
    });
  });
  return options;
}

export function getVisibleGroupLabels(groups: SelectData['groups']): Option[] {
  return groups.map((group) => group.options[0]).filter((option) => option && option.label && option.visible);
}

export function getSelectedOptions(groups: SelectData['groups']): Option[] {
  return getAllOptions(groups).filter((option) => !!option.selected);
}

export function validateOption(option: OptionInProps | string): Option {
  if (typeof option === 'string') {
    return { value: option, label: option, selected: false, isGroupLabel: false, visible: true, disabled: false };
  }

  const label = option.label || option.value || '';
  const value = option.value || label;

  return {
    label,
    value,
    selected: !!option.selected,
    isGroupLabel: false,
    visible: typeof option.visible === 'boolean' ? option.visible : true,
    disabled: typeof option.disabled === 'boolean' ? option.disabled : false,
  };
}

export function createGroupLabel(label: string) {
  return { ...validateOption(label), isGroupLabel: true, visible: !!label, disabled: false };
}

export function propsToGroups(props: Pick<SelectProps, 'groups' | 'options'>): SelectData['groups'] | undefined {
  if (!props.groups && !props.options) {
    return undefined;
  }
  if (props.groups) {
    return (props.groups as Group[]).map((group: Group) => {
      const hasLabelOptionAlready = !!group.options[0].isGroupLabel;
      const groupOptions = group.options.map(validateOption) as Option[];
      if (hasLabelOptionAlready) {
        return {
          options: [group.options[0], ...groupOptions.slice(1)],
        };
      }
      const labelOption: Option = createGroupLabel((group as GroupInProps).label);

      return {
        options: [labelOption, ...groupOptions],
      };
    });
  }

  return [
    {
      options: [createGroupLabel(''), ...(props.options || []).map(validateOption)],
    },
  ];
}

export function defaultFilter(option: Option, filterStr: string) {
  return option.label.toLowerCase().indexOf(filterStr.toLowerCase()) > -1;
}

export function filterOptions(groups: SelectData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function childrenToGroups(children: SelectProps['children']): SelectData['groups'] | undefined {
  if (!children || typeof children !== 'object') {
    return undefined;
  }
  const childArray = getChildrenAsArray(children) as ReactElement[];
  if (!childArray.length) {
    return [{ options: [] }];
  }
  const hasOptionGroups = childArray[0].type === 'optgroup';
  const optionElementToOption = (optionEl: ReactNode | ReactElement) => {
    const props = (
      optionEl && typeof optionEl === 'object' ? (optionEl as ReactElement).props : {}
    ) as ReactElement<HTMLOptionElement>['props'];
    const label = String(props.children);
    const value = props && String(props.value);
    const selected = !!(props && props.selected);
    const disabled = !!(props && props.disabled);
    return validateOption({ label, value, selected, disabled });
  };
  if (hasOptionGroups) {
    return childArray.map((child) => {
      const optionElements = child.props.children;
      const options = optionElements ? getChildrenAsArray(optionElements).map(optionElementToOption) : [];
      const label = createGroupLabel(String(child.props.label));
      options.unshift(label);
      return { options };
    });
  }
  return [{ options: [createGroupLabel(''), ...childArray.map(optionElementToOption)] }];
}

export function createSelectedOptionsList(currentSelections: Option[], groups: SelectData['groups']) {
  const selections = getSelectedOptions(groups);
  const selectionsAsValues = new Set(selections.map((opt) => opt.value));
  const stillSelected = currentSelections.filter((opt) => selectionsAsValues.has(opt.value));
  const selectedValues = new Set(stillSelected.map((opt) => opt.value));
  return [...stillSelected, ...selections.filter((opt) => !selectedValues.has(opt.value))];
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    arrowButton: `${containerId}-arrow-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function getGroupLabelOption(group: Group): Option | undefined {
  const firstOption = group.options[0];
  return firstOption && firstOption.isGroupLabel ? firstOption : undefined;
}
