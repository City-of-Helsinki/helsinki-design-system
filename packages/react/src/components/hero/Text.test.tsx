import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Hero } from './Hero';
import { getElementAttributesMisMatches, getCommonElementTestProps } from '../../utils/testHelpers';

describe('<Hero.Text /> spec', () => {
  const paragraphProps = getCommonElementTestProps('p');
  // axe will complain if tabIndex > 0
  paragraphProps.tabIndex = 0;
  const Component = () => {
    return (
      <Hero.Text {...paragraphProps}>
        Nullam ut nunc consectetur, accumsan nunc sed, luctus nisl. Curabitur lacinia!
      </Hero.Text>
    );
  };

  it('renders the component', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<Component />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('Native html props are passed to the element', async () => {
    const { getByTestId } = render(<Component />);
    const element = getByTestId(paragraphProps['data-testid']);
    expect(getElementAttributesMisMatches(element, paragraphProps)).toHaveLength(0);
  });
});
