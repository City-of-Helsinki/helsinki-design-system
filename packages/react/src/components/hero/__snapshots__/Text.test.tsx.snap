// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hero.Text /> spec renders the component 1`] = `
<DocumentFragment>
  <p
    aria-describedby="aria-describedby"
    aria-label="aria-label"
    class="text test-class-name"
    data-hds-prop="hdsProp"
    data-testid="testId"
    id="element-id"
    style="padding: 100px;"
    tabindex="0"
  >
    Nullam ut nunc consectetur, accumsan nunc sed, luctus nisl. Curabitur lacinia!
  </p>
</DocumentFragment>
`;
